dot/network/host.go

```diff
@@ -65,15 +63,27 @@ func newHost(ctx context.Context, cfg *Config) (*host, error) {
 	}
 
 	var externalAddr ma.Multiaddr
-	ip, err := pubip.Get()
-	if err != nil {
-		logger.Errorf("failed to get public IP: %s", err)
-	} else {
-		logger.Debugf("got public IP %s", ip)
+	if cfg.PublicIP != "" {
+		ip := net.ParseIP(cfg.PublicIP)
+		if ip == nil {
+			return nil, fmt.Errorf("invalid public ip: %s", cfg.PublicIP)
+		}
+		logger.Debugf("using config PublicIP: %s", ip)
 		externalAddr, err = ma.NewMultiaddr(fmt.Sprintf("/ip4/%s/tcp/%d", ip, cfg.Port))
 		if err != nil {
 			return nil, err
 		}
+	} else {
+		ip, err := pubip.Get()
+		if err != nil {
+			logger.Errorf("failed to get public IP error: %v", err)
+		} else {
+			logger.Debugf("got public IP", "IP", ip)
+			externalAddr, err = ma.NewMultiaddr(fmt.Sprintf("/ip4/%s/tcp/%d", ip, cfg.Port))
+			if err != nil {
+				return nil, err
+			}
+		}
 	}
```

Comment:
We could use if, else to just set ip and shift some common code out side those blocks

```go
	var ip net.IP
	if cfg.PublicIP != "" {
		ip = net.ParseIP(cfg.PublicIP)
		if ip == nil {
			return nil, fmt.Errorf("invalid public ip: %s", cfg.PublicIP)
		}
		logger.Debugf("using config PublicIP: %s", ip)

	} else {
		ip, err = pubip.Get()
		if err != nil {
			return nil, fmt.Errorf("failed to get public IP error: %v", err)
		}
		logger.Debugf("got public IP", "IP", ip)
	}

	externalAddr, err := ma.NewMultiaddr(fmt.Sprintf("/ip4/%s/tcp/%d", ip, cfg.Port))
	if err != nil {
		return nil, err
	}
```
(Looks like, code suggestions are not supported for more that a few lines)

Reply:
I'm adding another flag for `cfg.PublicDNS` so this is changing in the next PR, to have an `else if` case where the multiaddr is no longer `"/ip4/%s.."`, but rather `"/dns/%s..."`.